presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.log.Logger;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.spi.type.Type;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Verify.verify;
+
+public class DynamicFilter
+{
+    private static final Logger log = Logger.get(DynamicFilter.class);
+
+    private final List<Descriptor> descriptors;
+    private final TableScanNode tableScanNode;
+    private final SettableFuture<TupleDomain<ColumnHandle>> resultFuture;
+
+    private final Consumer<TupleDomain<ColumnHandle>> tupleDomainConsumer;
+    private TupleDomain<ColumnHandle>[] partitions;
+    private AtomicInteger partitionCounter;
+
+    public DynamicFilter(List<Descriptor> descriptors, TableScanNode tableScanNode, int partitionCount)
+    {
+        this.descriptors = descriptors;
+        this.tableScanNode = tableScanNode;
+        this.resultFuture = SettableFuture.create();
+
+        this.partitions = new TupleDomain[partitionCount];
+        this.partitionCounter = new AtomicInteger();
+        this.tupleDomainConsumer = tupleDomain -> {
```

Comment:
extract that to separate method outside of constructor

Reply:
Done at https://github.com/prestosql/presto/commit/96a591f3652533cd2e2795d349556158e2b86b6b#diff-d3d099fe2c1d2a89408b9234c0d7ab29R99.